File: Problems & Solutions/2256. Minimum Average Difference/04 December 2022/Approach 3 - Prefix Sum Optimized/Minimum Average Difference.js
let minimumAverageDifference = function(nums) {
    let n = nums.length;
    let ans = -1;
    let minAvgDiff = Number.MAX_SAFE_INTEGER;
    let currPrefixSum = 0;
        
    let totalSum = 0;
    for (let index = 0; index < n; ++index) {
        totalSum += nums[index];
    }
    
    for (let index = 0; index < n; ++index) {
        currPrefixSum += nums[index];
        
        let leftPartAverage = currPrefixSum;
        leftPartAverage = Math.floor(leftPartAverage / (index + 1));

        let rightPartAverage = totalSum - currPrefixSum;
        if (index != n - 1) {
            rightPartAverage = Math.floor(rightPartAverage / (n - index - 1));
        }

        let currDifference = Math.abs(leftPartAverage - rightPartAverage);
        if (currDifference < minAvgDiff) {
            minAvgDiff = currDifference;
            ans = index;
        }
    }

    return ans;
};

File: Problems & Solutions/2256. Minimum Average Difference/04 December 2022/Approach 3 - Prefix Sum Optimized/Minimum Average Difference.py
class Solution:
    def minimumAverageDifference(self, nums: List[int]) -> int:
        n = len(nums)
        ans = -1
        min_avg_diff = math.inf
        curr_prefix_sum = 0
        
        # Get total sum of array.
        total_sum = 0
        for index in range(n):
            total_sum += nums[index]
        
        for index in range(n):
            curr_prefix_sum += nums[index]
            
            left_part_average = curr_prefix_sum
            left_part_average //= (index + 1)
            
            right_part_average = total_sum - curr_prefix_sum
            if index != n - 1:
                right_part_average //= (n - index - 1)
            
            curr_difference = abs(left_part_average - right_part_average)
            
            if curr_difference < min_avg_diff:
                min_avg_diff = curr_difference
                ans = index
                
        return ans